docs/Hosting.md

```diff
@@ -0,0 +1,417 @@
+* Bicep schema lookup (served from embedded manifest, not live ARM call).
+* Best practices guidance (static text file resource already present).
+* Tool listing / capabilities (subset) for UI validation.
+
+NOT allowed in demo mode (initial):
+* Any command requiring subscription, resource group, or tenant parameters.
+* Commands with network calls to Azure management/data plane.
+* Destructive or mutation operations.
+
+Implementation Sketch:
+1. Add mode detector early in startup; when active, replace tool registry with curated static tool set.
+2. Inject a response decorator adding demo warning metadata.
+3. Block any attempt to pass subscription/resource parameters (validation layer).
+4. Enforce rate limiting middleware (in-process token bucket) scoped to remote transport only.
+5. Auto-disable if an auth credential becomes available (optional; could prompt to restart in full mode).
+
+Risk Considerations:
+* Users may believe they are interacting with real Azure resources. → Mitigation: explicit labeling, banner, distinct tool names (e.g., `demo-storage-example-list`).
+* Attackers could scrape static data for amplification. → Mitigation: minimal static set, rate limiting.
+* Accidental production enablement. → Mitigation: require BOTH flag and env var (defense in depth) OR build-time compile symbol for distribution images.
+
+### 6.8 Ideal Entra ID (Authorization Code + PKCE + OBO) Flow & Gaps
+Target future architecture for per-user delegated access:
+
+**Target Sequence**
+1. MCP server advertises (in its Protocol Registration / PRM document) that it uses Microsoft Entra ID for auth (includes issuer / audience metadata pointers or discovery hints).
+2. Client connects to MCP server and retrieves the PRM document.
+3. Client discovers Entra as the Authorization Server (AS) from PRM metadata.
+4. Client constructs Entra OpenID Connect / OAuth2 metadata URLs (well-known discovery) and fetches endpoints.
+5. Client registers with the Authorization Server (dynamic client registration) OR already possesses a suitable client ID.
+6. Client initiates Authorization Code Flow with PKCE directly against Entra (user authenticates + consents).
+7. Entra issues an access token whose `aud` (audience) is the MCP server (plus optional ID token for client state).
+8. Client sends that access token to the MCP server on each request (e.g., Bearer token header / agreed channel field).
+9. MCP server validates token (issuer, signature, expiry, audience, scopes / roles, nonce / c_hash as needed).
+10. MCP server performs OAuth 2.0 On-Behalf-Of (OBO) flow to exchange the user token for downstream resource tokens (ARM, Storage, etc.).
+11. MCP server maps/associates downstream tokens to the inbound user principal (cache keyed by user object ID + scopes) and executes tool operations.
+12. Responses returned; auditing & telemetry capture per-user context.
+
+**Current Blocking Gap**: No dynamic client registration for arbitrary third-party native MCP clients.
```

Comment:
This is actually going to be your biggest problem. Because Entra ID doesn't support DCR, you will need to either (1) implement the AS part yourself for this, which puts you on the path of a token factory, or (2) ask every user to enter a client ID, which is, unfortunately, a pretty bad experience usability-wise.

Reply:
Den - Just so we know what we are saying no or yes to, could you let us know what it would mean to implement AS?

Reply:
AS = Authorization Server that understands DCR; basically roll our own security sensitive code for the auth server; so that clients could dynamically register or unregister themselves (GET,PUT,DELETE /register), receive their {unique_client_id, access_token, mgmt_uris} to then begin using standard oauth with Entra ID